### ISSUE TITLE
Logged results represent the last, not the best epoch

### ISSUE THREAD

**Issue by jveitchmichaelis** [ultralytics/yolov5#4858]
I've noticed that after training a model, the final reported metrics do not reflect true model performance. Validation appears to only be run on the `last` epoch, not the `best`. This typically under-represents model performance and is an issue for e.g. wandb logging because those are the numbers that get reported in the dashboard - this also affects sweeps because that's what wandb uses for e.g. Bayesian optimisation as well as the sweep chart. 

For example, here is an epoch that is near the best during a run (not sure exactly due to the weighting of the different mAP metrics):

```
  Epoch   gpu_mem       box       obj       cls    labels  img_size
     35/99     6.88G   0.01216   0.00235  0.000924        30       640: 100%|████████| 288/288 [04:09<00:00,  1.15it/s]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|█| 41/41 [00:29<00:00,  1.
                 all       2572      10936      0.446      0.467      0.454       0.18

...

```

However the  final output only reports values for last epoch - note the best all-mAP is reported as 0.388 / 0.14. Both values are lower than the best epoch values of ~0.454 / 0.18. This is also the same in the confusion matrix/PR curves which have 0.388 as the mAP value.

```
 Epoch   gpu_mem       box       obj       cls    labels  img_size
     99/99     6.88G  0.007745  0.001915 0.0002389        30       640: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 288/288 [04:09<00:00,  1.16it/s]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 41/41 [00:39<00:00,  1.05it/s]
                 all       2572      10936      0.449      0.361      0.388       0.14
                 A       2572       1751      0.826      0.756      0.818       0.36
                B       2572       5187      0.323      0.189       0.19     0.0482
               C       2572       3365       0.63      0.477      0.537      0.151
               D       2572        633     0.0187     0.0221    0.00579    0.00134
               
100 epochs completed in 8.410 hours.
Optimizer stripped from train_noaug\weights\last.pt, 14.4MB
Optimizer stripped from train_noaug\weights\best.pt, 14.4MB

wandb: Waiting for W&B process to finish, PID 2540
wandb: Program ended successfully.
wandb: Find user logs for this run at: E:\yolov5\wandb\offline-run-20210919_213721-1iwmf7vc\logs\debug.log
wandb: Find internal logs for this run at: E:\yolov5\wandb\offline-run-20210919_213721-1iwmf7vc\logs\debug-internal.log
wandb: Run summary:
wandb:                 train/box_loss 0.00774
wandb:                 train/obj_loss 0.00191
wandb:                 train/cls_loss 0.00024
wandb:              metrics/precision 0.4494
wandb:                 metrics/recall 0.36094
wandb:                metrics/mAP_0.5 0.38779
wandb:           metrics/mAP_0.5:0.95 0.14007
wandb:                   val/box_loss 0.02515
wandb:                   val/obj_loss 0.00545
wandb:                   val/cls_loss 0.00905
wandb:                          x/lr0 0.0001
wandb:                          x/lr1 0.0001
wandb:                          x/lr2 0.0001
wandb:                       _runtime 30313
wandb:                     _timestamp 1632074555
wandb:                          _step 100
```

In `train.py` post-training validation is run with `model = ema.ema` whereas if I run `val.py` with `best.pt`, I see significantly better results. I see no reason why you'd want to validate on the last epoch rather than the model that performs best.

Here's the output from `val.py` showing significantly better (and more representative results);

```
val: Scanning '..\splits\val_5_8bit.cache' images and labels... 2572 found, 0 missing, 0 empty, 0 corrupted: 100%|███████████████████████████████████████████████████████████████████████████████████████| 2572/2572 [00:00<?, ?it/s]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 81/81 [00:38<00:00,  2.11it/s]
                 all       2572      10936      0.439      0.422      0.487      0.198
                 A       2572       1751      0.775      0.855      0.879      0.457
                B       2572       5187          0          0      0.244     0.0635
               C       2572       3365      0.537      0.705      0.685      0.241
               D       2572        633      0.443       0.13       0.14     0.0294
Speed: 0.8ms pre-process, 6.0ms inference, 1.8ms NMS per image at shape (32, 3, 640, 640)
```

I'd suggest running validation with something like:

```
if not noval or final_epoch:  # Calculate mAP
                results, maps, _ = val.run(data_dict,
                                           batch_size=batch_size // WORLD_SIZE * 2,
                                           imgsz=imgsz,
                                           weights=None if not final_epoch else best,
                                           model=ema.ema if not final_epoch else None,
```

So normal validation during training is preserved, but the final one runs on the best weights - not the last model.

**Comment by glenn-jocher** [comment]
@jveitchmichaelis yes I think I understand. To clarify, this is a W&B specific issue you are commenting on right? 

@AyushExel is our W&B expert, I think he might be able to comment further. Ayush the main issue seems to be that W&B is logging final metrics for the run rather than best metrics in the dashboard (upon training completion). During in-progress training I think we do want the last results, but once training is complete the most relevant results are the best epoch values. This is difficult in practice because the best fitness epoch is not tracked anywhere once it's passed.

**Comment by AyushExel** [comment]
@glenn-jocher @jveitchmichaelis we can enable this. There's a simple fix which will log the max. Value of accuracy and min. value of losses in the summary metrics section. I'll make a PR soon and tag it here

**Comment by jveitchmichaelis** [comment]
Hi @glenn-jocher it's both I think - what gets printed at the end of a non-wandb run is also the final epoch, not the best - since the validation run is passed `ema.ema`. If you run val with `best` weights, you get different (usually better) results.

**Comment by glenn-jocher** [comment]
@jveitchmichaelis oh got it! Yes so you are talking about updating this section to run val.py on best.pt after a training run finishes. That makes sense. Right now we do this only for COCO for last and best.

https://github.com/ultralytics/yolov5/blob/59aae85a7e40701bb872df673a6ef288e99a4ae3/train.py#L407-L419

**Comment by glenn-jocher** [comment]
@jveitchmichaelis one thing to keep in mind is best.pt may not exist under some scenarios, i.e. when `--nosave` is used

**Comment by jveitchmichaelis** [comment]
@glenn-jocher I think the fall back to last.pt is sensible? That's done
elsewhere if best.pt doesn't exist. I had a play around, but I couldn't see
a simple way (e.g. memory efficient) to update the model that gets passed.

The way I suggested works in testing, e.g. calling `val` with either the
model ema or the specified weights file, but if you pass a weight file then
the validator instantiates a new model instance and if you don't have much
GPU RAM (for example you're at the limit of batch size) that can fail...

**Comment by jveitchmichaelis** [comment]
Btw i think it's actually here that the issue arises (where we only log the last epoch): https://github.com/ultralytics/yolov5/blob/59aae85a7e40701bb872df673a6ef288e99a4ae3/train.py#L351-L363

(since I'm not training on COCO and there is still a val output at the end of training)

**Comment by glenn-jocher** [comment]
@jveitchmichaelis ah, this is the per epoch val run, which is on by default unless --noval is used. This must run on the current model (`ema.ema`) to get the epoch metrics.

**Comment by jveitchmichaelis** [comment]
@glenn-jocher Agreed it needs to run per-epoch, but on the last epoch it also produces "hard" outputs like plots and the results JSON:

```
save_json=is_coco and final_epoch,
verbose=nc < 50 and final_epoch, 
plots=plots and final_epoch,
```

The plotting is probably a more serious problem imo, because I would assume (as a user) that those represent the best results. For example the auto-generated confusion matrix is most likely showing an underperforming model compared to `best.pt`.

**Comment by jveitchmichaelis** [comment]
Yeah I agree we need to run this per-epoch. How about...

```
     results, maps, _ = val.run(data_dict, 
                                batch_size=batch_size // WORLD_SIZE * 2, 
                                imgsz=imgsz, 
                                model=ema.ema, 
                                single_cls=single_cls, 
                                dataloader=val_loader, 
                                save_dir=save_dir, 
                                save_json=False, 
                                verbose=False, 
                                plots=False, 
                                callbacks=callbacks, 
                                compute_loss=compute_loss) 
```

and then:

```
         for m in [last, best] if best.exists() else [last]:  # speed, mAP tests 
             results, _, _ = val.run(data_dict, 
                                     batch_size=batch_size // WORLD_SIZE * 2, 
                                     imgsz=imgsz, 
                                     model=attempt_load(m, device).half(), 
                                     iou_thres=0.7,  # NMS IoU threshold for best pycocotools results 
                                     single_cls=single_cls, 
                                     dataloader=val_loader, 
                                     save_dir=save_dir, 
                                     verbose=nc < 50,
                                     save_json=is_coco, 
                                     plots=plots and not is_coco) 
```

This preserves existing functionality (and behaviour), but moves the validation to the second loop.

**Comment by jveitchmichaelis** [comment]
Yeah I agree we need to run this per-epoch. How about...

Per Epoch:
```
     results, maps, _ = val.run(data_dict, 
                                batch_size=batch_size // WORLD_SIZE * 2, 
                                imgsz=imgsz, 
                                model=ema.ema, 
                                single_cls=single_cls, 
                                dataloader=val_loader, 
                                save_dir=save_dir, 
                                save_json=False, 
                                verbose=False, 
                                plots=False, 
                                callbacks=callbacks, 
                                compute_loss=compute_loss) 
```

and then post training:

```
         for m in [last, best] if best.exists() else [last]:  # speed, mAP tests 
             results, _, _ = val.run(data_dict, 
                                     batch_size=batch_size // WORLD_SIZE * 2, 
                                     imgsz=imgsz, 
                                     model=attempt_load(m, device).half(), 
                                     iou_thres=0.7,  # NMS IoU threshold for best pycocotools results 
                                     single_cls=single_cls, 
                                     dataloader=val_loader, 
                                     save_dir=save_dir, 
                                     verbose=nc < 50,
                                     save_json=is_coco, 
                                     plots=plots and not is_coco) 
```

This preserves existing functionality (and behaviour), but moves the validation to the second loop.

**Comment by glenn-jocher** [comment]
@jveitchmichaelis yes that makes sense. So I think we want an extra val call at the end of training to run on best.pt with plots turned on:

```
train epoch 0
val epoch 0
train epoch 1
val epoch 1
...
train epoch 299
val epoch 299

val epoch best --plots True
```

**Comment by glenn-jocher** [comment]
@jveitchmichaelis about the CUDA usage, I'd be surprised in an extra val.py call with --weights best.pt would increase CUDA usage, as most of the usage is from backwards calls on the losses I think during training. When val.py runs there should be quite low CUDA utilization (before, during and after).

**Comment by glenn-jocher** [comment]
@jveitchmichaelis @AyushExel see PR #4889 for a possible fix. This PR adds an additional val step at the end of training on best.pt (if it exists). 

One issue is that if best.pt doesn't exist (--nosave or --noval) then no plots will be saved... hmm.

**Comment by AyushExel** [comment]
@glenn-jocher the additional val step at the end should also log the metrics to W&B, right?

**Comment by glenn-jocher** [comment]
@AyushExel yes that's the idea. The alternative is to save plots and confusion matrix etc. on every val.py run, but these extra outputs add processing time to the val.py run. I'm not sure what the right solution is.

**Comment by jveitchmichaelis** [comment]
One option is to only save if it's a best run? But in general I don't think
we need plots every epoch, at the end of training is fine . How long does
plotting take versus running val? I'm guessing not a huge amount compared
to running inference on a val set.

**Comment by glenn-jocher** [comment]
@jveitchmichaelis the plots are not fast. Profiling results on Colab:
```python
%timeit !python val.py --weights yolov5s.pt --data coco128.yaml --img 640 --iou 0.65 --half
```

```python
1 loop, best of 5: 16.1 s per loop  # plots=True
1 loop, best of 5: 7.65 s per loop  # plots=False
```

EDIT: additional overhead is not just plotting, but computing values for confusion matrix etc.

**Comment by glenn-jocher** [comment]
@jveitchmichaelis @AyushExel good news 😃! Your original issue may now be fixed ✅ in PR #4889. This PR runs validation on best.pt after training completes, having two main effects: 1) verbose mode is on so users see per-class metrics now automatically on training completion, 2) plots are turned on so plots and images use best.pt metrics rather than last.pt metrics as before. @AyushExel in my tests everything seemed to work well with W&B but please take a look yourself also!

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by jveitchmichaelis** [comment]
Great, thanks @glenn-jocher - I'll try it out!
